exercises/handling-errors/practice/src/activities.ts

```diff
@@ -0,0 +1,77 @@
+// TODO Part A: Import `ApplicationFailure` from `@temporalio/common`
+import { Address, Bill, Distance, OrderConfirmation } from './shared';
+import { log } from '@temporalio/activity';
+
+export async function getDistance(address: Address): Promise<Distance> {
+  log.info('getDistance invoked; determining distance to customer address');
+
+  // this is a simulation, which calculates a fake (but consistent)
+  // distance for a customer address based on its length. The value
+  // will therefore be different when called with different addresses,
+  // but will be the same across all invocations with the same address.
+  let kilometers: number = address.line1.length + address.line2.length - 10;
+  if (kilometers < 1) {
+    kilometers = 5;
+  }
+
+  const distance = {
+    kilometers,
+  };
+
+  log.info('getDistance complete', { distance });
+  return distance;
+}
+
+export async function sendBill(bill: Bill): Promise<OrderConfirmation> {
+  log.info('sendBill invoked', { Customer: bill.customerID, Amount: bill.amount });
+
+  let chargeAmount = bill.amount;
+
+  // This month's special offer: Get $5 off all orders over $30
+  if (bill.amount > 3000) {
+    log.info('Applying discount');
+
+    chargeAmount -= 500; // reduce amount charged by 500 cents
+  }
+
+  // reject invalid amounts before calling the payment processor
+  if (chargeAmount < 0) {
+    throw ApplicationFailure.create({
+      message: `Invalid charge amount: ${chargeAmount} (must be above zero)`,
+      details: [chargeAmount],
+      // TODO Part B: Set the nonRetryable key to true.
+    });
+  }
+
+  // pretend we called a payment processing service here :-)
+
+  const confirmation = {
+    orderNumber: bill.orderNumber,
+    confirmationNumber: 'AB9923',
+    status: 'SUCCESS',
+    billingTimestamp: Date.now(),
+    amount: chargeAmount,
+  };
+
+  log.info('sendBill complete', { confirmation });
+
+  return confirmation;
+}
+
+export async function validateCreditCard(creditCardNumber: string): Promise<void> {
+  log.info('validateCreditCard invoked', { CreditCardNumber: creditCardNumber });
+
+  // Check if the credit card number has 16 digits
+  const isValid = creditCardNumber.length == 16;
+
+  if (!isValid) {
+    // TODO Part A: Replace the log statement below
+    // Throw an `ApplicationFailure` if the
+    // credit card number does not have 16 digits.
+    // Follow the pattern you see for in the error in `sendBill`.
+    // TODO Part B: Set the nonRetryable key to true.
+    log.error('invalid credit card number');
```

Comment:
Nit: Log messages and print statements in samples should be subject to the same formality that developers use to communicate in other ways. Recommend uppercasing.

Nit: I'd reverse the logic and make the term more readable. Instead of `isValid`, which must be checked by negation, I'd use `invalidCreditCardNumber = creditCardNumber.length != 16;`